cardano-crypto-class/src/Cardano/Crypto/SECP256K1/C.hs

```diff
@@ -74,5 +111,75 @@ foreign import capi "secp256k1_schnorrsig.h secp256k1_schnorrsig_verify"
   -> SizedPtr SECP256K1_SCHNORR_SIGNATURE_BYTES -- signature to verify (64 bytes)
   -> Ptr CUChar -- message to verify
   -> CSize -- message length in bytes
-  -> SizedPtr SECP256K1_XONLY_PUBKEY_BYTES -- pubkey to verify with
+  -> SizedPtr SECP256K1_SCHNORR_PUBKEY_BYTES_INTERNAL -- pubkey to verify with
   -> CInt -- 1 on success, 0 on failure
+
+foreign import capi "secp256k1_extrakeys.h secp256k1_xonly_pubkey_serialize"
+  secpXOnlyPubkeySerialize :: 
+     Ptr SECP256k1Context -- an initialized context
+  -> SizedPtr SECP256K1_SCHNORR_PUBKEY_BYTES -- out-param for serialized representation
+  -> SizedPtr SECP256K1_SCHNORR_PUBKEY_BYTES_INTERNAL -- the xonly pubkey to serialize
+  -> IO CInt -- 1 on success, 0 on error
+
+foreign import capi "secp256k1_extrakeys.h secp256k1_xonly_pubkey_parse"
+  secpXOnlyPubkeyParse ::
+     Ptr SECP256k1Context -- an initialized context
+  -> SizedPtr SECP256K1_SCHNORR_PUBKEY_BYTES_INTERNAL -- out-param for deserialized representation
+  -> Ptr Word8 -- bytes to deserialize
+  -> IO CInt -- 1 if the parse succeeded, 0 if the parse failed (due to invalid representation)
+
+foreign import capi "secp256k1.h secp256k1_ec_pubkey_create" 
+  secpEcPubkeyCreate :: 
+     Ptr SECP256k1Context -- an initialized context
+  -> SizedPtr SECP256K1_ECDSA_PUBKEY_BYTES_INTERNAL -- out-param for generated key
+  -> SizedPtr SECP256K1_ECDSA_PRIVKEY_BYTES -- seed private key
+  -> IO CInt -- 1 on success, 0 on error
+
+foreign import capi "secp256k1.h secp256k1_ecdsa_sign"
+  secpEcdsaSign :: 
+     Ptr SECP256k1Context -- context initialized for signing
+  -> SizedPtr SECP256K1_ECDSA_SIGNATURE_BYTES_INTERNAL -- out-param for signature
+  -> SizedPtr 32 -- pointer to 32 bytes of hashed message data
+  -> SizedPtr SECP256K1_ECDSA_PRIVKEY_BYTES -- private key to sign with
+  -> Ptr CChar -- not used
+  -> Ptr CChar -- not used
```

Comment:
Can we do something silly that can only be inhabited by a null pointer like `Ptr Void`?

Reply:
This is not really a valid protection for pointers those bastards get cast all the time  :smile:. Even in this PR there are a few `castPtr`

Reply:
Yeah, Haskellers should really use `Ptr Any` instead of `Ptr Void` :smile:

Reply:
I would have thought that `void` in C is quite different from `Void` in Haskell, so I'm surprised they correspond here :thinking:

Reply:
Of course it is different, with some similarities:
* it has same name
* value of type void can't be constructed

On the other hand when used as `void*` in C it usually means a pointer to a buffer with values of statically unknown type, but there very well could be data in that buffer that the pointer is referring to, unless it is null of course. Which more corresponds to a `Ptr Any` in Haskell IMHO, rather than `Ptr Void`. It is quite common to use `Ptr ()` for this as well. But considering that neither one of those choices gives us any sort of type safety this discussion is totally irrelevant.

Reply:
This is a common Cism - something can mean totally different things when used in a different context. In the context of a return from a function, `void` means 'no value'. In the context of a _pointer_, `void` means 'unknown/any value'. This is also how `0` in a pointer context means `NULL`, even though the standard doesn't require `NULL` to be address 0 - merely distinct from any non-`NULL` address.